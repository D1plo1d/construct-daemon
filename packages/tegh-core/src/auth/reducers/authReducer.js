import { loop, Cmd } from 'redux-loop'
import { Map, Record } from 'immutable'
import { ec as EC } from 'elliptic'

import displayInviteInConsole from '../sideEffects/displayInviteInConsole'

import { SET_CONFIG } from '../../config/actions/setConfig'

import createInvite from '../actions/createInvite'
import { DAT_PEER_HANDSHAKE_RECEIVED } from '../actions/datPeerHandshakeReceived'
import { DAT_PEER_DATA_RECEIVED } from '../actions/datPeerDataReceived'

const ec = new EC('curve25519')

export const initialState = Record({
  hostKeys: null,
  usersAndInvitesByPublicKey: Map(),
  datSessionsByPeerID: Map(),
})

const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = action.payload
      const { users, invites } = config.auth

      const admins = users.filter(user => user.admin)
      const adminInvites = invites.filter(invite => invite.admin)
      let consoleInvites = invites.filter(invite => invite.displayInConsole)

      const sideEffects = []

      /*
       * If no invites exist and no admins exist then create an invite so the
       * initial user can connect
       */
      if (admins.size === 0 && adminInvites.size === 0) {
        const nextAction = createInvite({
          displayInConsole: true,
          admin: true,
        })
        consoleInvites = consoleInvites.push(nextAction.payload.invite)
        sideEffects.push(Cmd.action(nextAction))
      }

      /*
       * Display invites in the console
       */
      consoleInvites.forEach(invite => (
        sideEffects.push(Cmd.run(displayInviteInConsole, {
          args: [{ hostPublicKey: hostKeys.getPublic(), invite }],
        }))
      ))

      return loop(state, Cmd.list(sideEffects))
    }
    case DAT_PEER_HANDSHAKE_RECEIVED: {
      const hostEphemeralKeys = ec.genKeyPair()
      hostEphemeralKeys.getPublic()
    }
    case DAT_PEER_DATA_RECEIVED: {
    }
    default: {
      return state
    }
  }
}

export default authReducer
