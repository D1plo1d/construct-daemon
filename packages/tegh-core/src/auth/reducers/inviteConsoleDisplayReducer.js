import { loop, Cmd } from 'redux-loop'
import { Map, Record } from 'immutable'

import displayInviteInConsole from '../sideEffects/displayInviteInConsole'

import { SET_CONFIG } from '../../config/actions/setConfig'

import createInvite from '../actions/createInvite'

export const initialState = Record({
  hostIdentityKeys: null,
  // accounts are users and invites mapped by their public key
  accounts: Map(),
})

const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = action.payload
      const {
        users,
        invites,
        hostIdentityKeys,
      } = config.auth

      const admins = users.filter(user => user.admin)
      const adminInvites = invites.filter(invite => invite.admin)
      let consoleInvites = invites.filter(invite => invite.displayInConsole)

      let sideEffects = []

      /*
       * If no invites exist and no admins exist then create an invite so the
       * initial user can connect
       */
      if (admins.size === 0 && adminInvites.size === 0) {
        const nextAction = createInvite({
          displayInConsole: true,
          admin: true,
        })
        consoleInvites = consoleInvites.push(nextAction.payload.invite)
        sideEffects.push(Cmd.action(nextAction))
      }

      /*
       * Display invites in the console
       */
      sideEffects = sideEffects.concat(consoleInvites.map(invite => (
        Cmd.run(displayInviteInConsole, {
          args: [{ hostPublicKey: hostIdentityKeys.getPublic(), invite }],
        })
      )))

      return loop(state, Cmd.list(sideEffects))
    }
    default: {
      return state
    }
  }
}

export default authReducer
