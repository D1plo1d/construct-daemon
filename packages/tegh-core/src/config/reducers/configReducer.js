import { loop, Cmd } from 'redux-loop'
import immutablePatch from 'immutablepatch'

import { SET_CONFIG } from '../actions/setConfig'
import requestSetConfig from '../actions/requestSetConfig'
import { REQUEST_PATCH_CONFIG } from '../actions/requestPatchConfig'

export const initialState = null

const configReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_PATCH_CONFIG: {
      const { patch } = action.payload

      const config = immutablePatch(state, patch)

      return loop(
        state,
        Cmd.action(requestSetConfig({ config })),
      )
    }
    case SET_CONFIG: {
      const { config } = action.payload

      return config
    }
    default: {
      return state
    }
  }
}

export default configReducer
