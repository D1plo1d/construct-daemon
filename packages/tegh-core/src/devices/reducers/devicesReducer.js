import { loop, Cmd } from 'redux-loop'
import { Map, Record } from 'immutable'

import watchSerialPorts from '../sideEffects/watchSerialPorts'

import { SET_CONFIG } from '../../config/actions/setConfig'
import { DEVICE_CONNECTED } from '../actions/deviceConnected'
import { DEVICE_DISCONNECTED } from '../actions/deviceDisconnected'

export const initialState = Record({
  initialized: false,
  byID: Map(),
})()

const devicesReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      if (state.initialized) return state

      const nextState = state.set('initialized', true)

      return loop(
        nextState,
        Cmd.run(watchSerialPorts, {
          args: [Cmd.getState, Cmd.dispatch],
        }),
      )
    }
    case DEVICE_CONNECTED: {
      const { device } = action.payload
      // console.log('CONNECT', device.id)

      return state.setIn(['byID', device.id], device)
    }
    case DEVICE_DISCONNECTED: {
      const { device } = action.payload
      // console.log('DISCONNECT', device.id)

      return state.removeIn(['byID', device.id])
    }
    default: {
      return state
    }
  }
}

export default devicesReducer
