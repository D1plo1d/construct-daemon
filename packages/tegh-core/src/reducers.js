export config from './config/reducers/configReducer'
export devices from './devices/reducers/devicesReducer'
export jobQueue from './jobQueue/reducers/jobQueueReducer'
export log from './log/reducers/logReducer'
export macros from './macros/reducers/macrosReducer'
export pluginManager from './pluginManager/reducers/pluginManagerReducer'
export spool from './spool/reducers/spoolReducer'
export status from './printer/reducers/statusReducer'
