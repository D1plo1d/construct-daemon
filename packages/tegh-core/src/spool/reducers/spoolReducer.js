import { loop, Cmd } from 'redux-loop'
import {
  Record, List, Map,
} from 'immutable'

import isIdle from '../selectors/isIdle'

import {
  EMERGENCY,
  NORMAL,
  PREEMPTIVE,
  priorityOrder,
} from '../types/PriorityEnum'

import {
  PRINTING,
  isSpooled,
} from '../types/TaskStatusEnum'

/* printer actions */
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'
/* job actions */
import { CANCEL_JOB } from '../../jobQueue/actions/cancelJob'
import { DELETE_JOB } from '../../jobQueue/actions/deleteJob'
/* task actions */
import { SPOOL_TASK } from '../actions/spoolTask'
import requestDespool, { REQUEST_DESPOOL } from '../actions/requestDespool'
import despoolTask from '../actions/despoolTask'
import { CANCEL_TASK } from '../actions/cancelTask'
import cancelAllTasks, { CANCEL_ALL_TASKS } from '../actions/cancelAllTasks'

export const initialState = Record({
  priorityQueues: Record({
    [EMERGENCY]: List(),
    [PREEMPTIVE]: List(),
    [NORMAL]: List(),
  })(),
  tasks: Map(),
  currentTaskID: null,
})()

const removeTaskReferences = (state) => {
  const nextTaskIDs = state.tasks.map(task => task.id)

  let nextState = state
    .update('priorityQueues', queues => queues.map(queue => (
      queue.filter(taskID => nextTaskIDs.includes(taskID))
    )))

  if (nextTaskIDs.includes(state.currentTaskID) === false) {
    nextState = nextState.set('currentTaskID', null)
  }

  return nextState
}

const spoolReducer = () => (state = initialState, action) => {
  switch (action.type) {
    /* Spool reset actions */
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR:
    case CANCEL_ALL_TASKS: {
      return initialState
    }
    case CANCEL_JOB:
    case DELETE_JOB: {
      const { jobID } = action.payload

      const nextState = state.update('tasks', tasks => (
        tasks.filter(task => task.jobID !== jobID)
      ))

      return removeTaskReferences(nextState)
    }
    case CANCEL_TASK: {
      const { taskID } = action.payload

      const nextState = state.update('tasks', tasks => tasks.remove(taskID))

      return removeTaskReferences(nextState)
    }
    case SPOOL_TASK: {
      const { payload } = action
      const { id, priority } = payload.task

      let nextState = state

      /*
      * if the task is an emergency then cancel all other tasks in the queue
      */
      if (priority === EMERGENCY) {
        nextState = initialState
      }

      nextState = state.setIn(['tasks', id], payload.task)

      if (
        isIdle.resultFunc(state.tasks) === false
        && priority !== EMERGENCY
        && payload.task.internal !== true
      ) {
        throw new Error('Cannot spool non-emergency tasks when printing a job')
      }


      /* create the task */
      const createAction = createTask({ task: payload.task })
      nextState = taskMap.createOne(nextState, createAction)

      const taskQueue = ['priorityQueues', priority]
      const shouldDespool = nextState.currentTaskID == null

      nextState = nextState
        .updateIn(taskQueue, list => list.push(id))

      /*
       * despool the first line if nothing is spooled
       */
      if (shouldDespool) {
        return loop(nextState, Cmd.action(requestDespool()))
      }

      return nextState
    }
    case REQUEST_DESPOOL: {
      const { priorityQueues, currentTaskID } = state
      let nextState = state
      if (currentTaskID != null) {
        /*
         * despool the next line or finish the task via the taskReducer
         */
        nextState = taskMap.updateOne(nextState, action, currentTaskID)
      }
      const currentTask = nextState.tasks.get(currentTaskID)
      if (currentTask == null || currentTask.status !== PRINTING) {
        /*
         * if the current task is done then despool the next task if there is
         * anything to despool.
         */
        const priority = priorityOrder.find(priorityOption => (
          priorityQueues[priorityOption].size > 0
        ))

        if (priority == null) {
          nextState = nextState.set('currentTaskID', null)
        } else {
          const nextTaskID = state.priorityQueues[priority].first()
          /*
           * start the task via the taskReducer
           */
          nextState = nextState
            .mergeIn(['tasks', nextTaskID], {
              startedAt: new Date().toISOString(),
              status: PRINTING,
              currentLineNumber: 0,
            })
            .set('currentTaskID', nextTaskID)
            .updateIn(['priorityQueues', priority], list => list.shift())
        }
      }

      if (nextState.currentTaskID == null) {
        return nextState
      }

      const nextTask = nextState.tasks.get(nextState.currentTaskID)
      return loop(nextState, Cmd.action(despoolTask(nextTask)))
    }
    case REQUEST_DESPOOL: {
      if (state.currentLineNumber < state.data.size - 1) {
        /*
         * if the task has more lines to execute then increment the line number
         */
        return state.update('currentLineNumber', i => i + 1)
      }
      /* mark tasks as done after they are completed */
      return state.merge({
        // TODO: stoppedAt should eventually be changed to be sent after
        // the printer sends 'ok' or 'error' and should be based off
        // estimated print time
        stoppedAt: new Date().toISOString(),
        status: DONE,
        /* the data of each completed task is deleted to save space */
        data: null,
      })
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
