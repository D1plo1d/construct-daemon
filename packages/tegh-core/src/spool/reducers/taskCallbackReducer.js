import { loop, Cmd } from 'redux-loop'
import { Record, List } from 'immutable'

// TODO: implement task errored in spoolReducer
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'

import { TASK_ERRORED } from '../actions/taskErrored'
import { DESPOOL_TASK } from '../actions/despoolTask'
import { DESPOOL_COMPLETED } from '../actions/despoolCompleted'

export const initialState = List()

const spoolReducer = (state = initialState, action) => {
  switch (action.type) {
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR: {
      return initialState
    }
    case TASK_ERRORED: {
      const { task } = action.payload

      if (task.onError) {
        return loop(
          state,
          Cmd.run(task.onError, {
            args: [task],
          }),
        )
      }

      return state
    }
    case DESPOOL_COMPLETED: {
      const { task, isLastLineInTask } = action.payload

      const nextState = state.shift()

      if (isLastLineInTask && task.onComplete) {
        return loop(
          nextState,
          Cmd.run(task.onComplete, {
            args: [task],
          }),
        )
      }

      return initialState
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
