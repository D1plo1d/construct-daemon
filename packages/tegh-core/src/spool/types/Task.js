import type { RecordOf, List as ListT } from 'immutable'
import uuid from 'uuid'
import { Record, List } from 'immutable'
import type { PriorityEnumT } from './PriorityEnum'
import type { TaskStatusEnumT } from './TaskStatusEnum'


import normalizeGCodeLines from '../../util/normalizeGCodeLines'
import { priorityOrder, NORMAL } from './PriorityEnum'
import { SPOOLED } from './TaskStatusEnum'

export type TaskT = RecordOf<{
  id: string,
  jobID: ?string,
  jobFileID: ?string,

  priority: PriorityEnumT,
  internal: boolean,
  data: ListT<string>,
  name: ?string,

  /*
   * The line number that will be executed by the printer by a saga after the
   * reducers run.
   */
  currentLineNumber: ?number,

  createdAt: ?number,
  startedAt: ?number,
  stoppedAt: ?number,
  status: TaskStatusEnumT,
}>

const TaskRecord = Record({
  id: null,
  priority: null,
  internal: null,
  name: null,
  jobID: null,
  jobFileID: null,
  data: null,
  status: null,
  currentLineNumber: null,
  createdAt: null,
  startedAt: null,
  stoppedAt: null,
})

const Task = (attrs) => {
  const {
    priority,
    internal,
    name,
    data,
  } = attrs

  if (typeof name !== 'string') {
    throw new Error('name must be a string')
  }
  if (typeof internal !== 'boolean') {
    throw new Error('internal must be a boolean')
  }
  if (!priorityOrder.includes(priority)) {
    throw new Error('invalid priority')
  }
  if (!Array.isArray(data)) {
    throw new Error('data must be an array of strings')
  }
  return TaskRecord({
    id: uuid.v4(),
    createdAt: new Date().toISOString(),
    status: SPOOLED,
    ...attrs,
    data: List(normalizeGCodeLines(data)),
  })
}

export const MockTask = attrs => Task({
  name: 'test.ngc',
  internal: true,
  priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
  ...attrs,
})

export default Task
