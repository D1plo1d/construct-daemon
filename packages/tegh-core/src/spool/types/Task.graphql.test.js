import snapshotTestGraphQLType from '../../util/testing/snapshotTestGraphQLType'

import { initialState as spoolState } from '../reducers/spoolReducer'

import Task from './Task'
import { NORMAL } from './PriorityEnum'

import TaskGraphQL from './Task.graphql.js'

const task = Task({
  name: 'test task',
  internal: false,
  priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
})

const state = {
  spool: spoolState.setIn(['tasks', task.id], task),
  config: {
    id: 'test_printer_id',
  },
}

snapshotTestGraphQLType('TaskGraphQL', {
  type: TaskGraphQL,
  fieldConfigs: {
    id: {
      dynamic: true,
    },
    createdAt: {
      dynamic: true,
    },
    percentComplete: {
      args: { digits: 5 },
    },
  },
  rootValue: task,
  contextValue: {
    store: {
      getState: () => state,
    },
  },
})
