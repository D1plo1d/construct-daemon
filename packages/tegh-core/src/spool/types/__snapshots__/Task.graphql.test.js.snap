// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskGraphQL presents a consistent API 1`] = `
Object {
  "description": "A spooled set of gcodes to be executed by the printer",
  "enumValues": null,
  "fields": Array [
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "id",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "ID",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "name",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "currentLineNumber",
      "type": Object {
        "kind": "SCALAR",
        "name": "Int",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "totalLineNumbers",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [
        Object {
          "defaultValue": null,
          "description": "The number of digits to the right of the decimal place to round to. eg. \`digits: 0\` => 83 \`digits: 1\` => 82.6 \`digits: 2\` => 82.62",
          "name": "digits",
          "type": Object {
            "kind": "NON_NULL",
            "name": null,
            "ofType": Object {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null,
            },
          },
        },
      ],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "percentComplete",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Float",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "createdAt",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Date",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "startedAt",
      "type": Object {
        "kind": "SCALAR",
        "name": "Date",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "stoppedAt",
      "type": Object {
        "kind": "SCALAR",
        "name": "Date",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "status",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "printer",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "OBJECT",
          "name": "Printer",
          "ofType": null,
        },
      },
    },
  ],
  "inputFields": null,
  "interfaces": Array [],
  "kind": "OBJECT",
  "name": "Task",
  "possibleTypes": null,
}
`;

exports[`TaskGraphQL resolves every field without error 1`] = `
Object {
  "createdAt": "[DYNAMIC FIELD OMITTED]",
  "currentLineNumber": null,
  "id": "[DYNAMIC FIELD OMITTED]",
  "name": "test task",
  "percentComplete": 0,
  "printer": Object {
    "id": "[DYNAMIC FIELD OMITTED]",
  },
  "startedAt": null,
  "status": "SPOOLED",
  "stoppedAt": null,
  "totalLineNumbers": 2,
}
`;
